kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -307,6 +307,38 @@ private static List<Tuple2<List<String>, StructField>> recurseIntoComplexTypes(
     return filtered;
   }
 
+  /* Compute the SchemaChanges using field IDs */
+  static SchemaChanges computeSchemaChangesById(
+      Map<Long, StructField> currentFieldIdToField, Map<Long, StructField> updatedFieldIdToField) {
+    SchemaChanges.Builder schemaDiff = SchemaChanges.builder();
+    for (Map.Entry<Long, StructField> fieldInUpdatedSchema : updatedFieldIdToField.entrySet()) {
+      StructField existingField = currentFieldIdToField.get(fieldInUpdatedSchema.getKey());
+      StructField updatedField = fieldInUpdatedSchema.getValue();
+      // New field added
+      if (existingField == null) {
+        schemaDiff.withAddedField(updatedField);
+      } else if (isUpdatedField(existingField, fieldInUpdatedSchema.getValue())) {
+        // Field renamed or changed types
+        schemaDiff.withUpdatedField(existingField, fieldInUpdatedSchema.getValue());
+      }
+    }
+
+    for (Map.Entry<Long, StructField> entry : currentFieldIdToField.entrySet()) {
+      StructField fieldInUpdatedSchema = updatedFieldIdToField.get(entry.getKey());
+      if (fieldInUpdatedSchema == null) {
```

Comment:
nit - this could probably be one line `if (updatedFieldIdToField.contains(entry.getKey()))`